app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataController.kt

```diff
@@ -39,6 +41,21 @@ class MapDataController internal constructor(
     }
     private val listeners: MutableList<Listener> = CopyOnWriteArrayList()
 
+    val spatialCache = SpatialCache(
+        80, // TODO: 192 is likely WAY too much for S4 mini, maybe make it depend on available memory?
```

Comment:
Given that you only edit locally, i.e. in the same ~1-4 tiles at a time,  a cache of 4 tiles seems to be enough. However, since you can also zoom out and look at data (you will not likely edit now), the reasonable upper limit should be the number of tiles that can be in view at one time at zoom level 14 because zoom level 14 is the minimum zoom level at which quest pins are displayed.

Reply:
Heh, but that would be about 6x12 tiles (z16) on my phone. Probably more on a tablet. 

I guess, one has to ask what would be the behavior when the cache is smaller than that.

Reply:
Hmm cache smaller than number of tiles requested will cause issues.

The spatial cache will load everything it gets from `fetch`, but then immediately drop some of the tiles. Since the other caches are filled in `fetch`, on dropping tiles some of the content will be removed.
The list of node ids returned by the spatial will be complete, but by the time `spatialCache.get(bbox)` is done, some parts of the other caches will be dropped already, and will not be added to the data.

Reply:
Most simple solution would be moving the call to `trim` to `getMapDataWithGeometry`, after `elementGeometries` and `elements` are created.

Reply:
> Hmm cache smaller than number of tiles requested will cause issues.

True, but that is much less than what can potentially be displayed. See `QuestPinsManager.onNewScreenPosition`:
```kotlin
        // area too big -> skip (performance)
        if (tilesRect.size > 16) return
```
Pins will only be retrieved if the tile rect is smaller or equal ~ 4x4 = 16 tiles. So I guess this makes 16 tiles the absolute minimum tile cache size.

Reply:
Oh right, so I can get away with just adding some comment that `maxTiles` needs to be >= 16 for this reason.

Reply:
I added comments about this, fixed problems if bbox is larger than fits in cache for `SpatialCache.replaceAllInBBox`, and added a crash that will happen if bbox of `SpatialCache.get` does not fit in cache (as this should never happen)